ecs-cli/modules/clients/aws/route53/client_test.go

```diff
@@ -0,0 +1,321 @@
+
+// Package servicediscovery contains functions for working with the route53 APIs
+// that back ECS Service Discovery
+package route53
+
+import (
+	"testing"
+
+	"github.com/aws/amazon-ecs-cli/ecs-cli/modules/config"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/route53"
+	"github.com/aws/aws-sdk-go/service/servicediscovery"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestFindPrivateNamespace(t *testing.T) {
+	setupNamespaceMocks(t, false)
+	setupHostedZoneMocks(t)
+
+	var testCases = []struct {
+		testName   string
+		name       string
+		vpc        string
+		region     string
+		expectedID *string
+	}{
+		{
+			testName:   "Find namespace1",
+			name:       "corp",
+			vpc:        "vpc-8BAADF00D",
+			region:     "us-east-1",
+			expectedID: aws.String("namespace1"),
+		},
+		{
+			testName:   "Find namespace2",
+			name:       "prod",
+			vpc:        "vpc-1CEB00DA",
+			region:     "us-east-1",
+			expectedID: aws.String("namespace2"),
+		},
+		{
+			testName:   "Find namespace4",
+			name:       "corp",
+			vpc:        "vpc-C00010FF",
+			region:     "sa-east-1",
+			expectedID: aws.String("namespace4"),
+		},
+		{
+			testName:   "Namespace Does Not Exist",
+			name:       "corp",
+			vpc:        "vpc-C00010FF",
+			region:     "us-east-1",
+			expectedID: nil,
+		},
+		{
+			testName:   "Namspace with multiple VPCs",
+			name:       "bridge",
+			vpc:        "vpc-D15EA5E",
+			region:     "ap-south-1",
+			expectedID: aws.String("namespace3"),
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.testName, func(t *testing.T) {
+			namespaceID, err := FindPrivateNamespace(testCase.name, testCase.vpc, setupParamsInTest(testCase.region))
+			assert.Equal(t, testCase.expectedID, namespaceID, "Expected namespace ID to match")
+			assert.NoError(t, err, "Unexpected error from FindPrivateNamespace")
+		})
+	}
+}
+
+func TestFindPublicNamespace(t *testing.T) {
+	setupNamespaceMocks(t, true)
+
+	var testCases = []struct {
+		testName   string
+		name       string
+		expectedID *string
+	}{
+		{
+			testName:   "Find namespace1",
+			name:       "corp",
+			expectedID: aws.String("namespace1"),
+		},
+		{
+			testName:   "Find namespace2",
+			name:       "prod",
+			expectedID: aws.String("namespace2"),
+		},
+		{
+			testName:   "Find namespace3",
+			name:       "bridge",
+			expectedID: aws.String("namespace3"),
+		},
+		{
+			testName:   "Namespace Does Not Exist",
+			name:       "cat",
+			expectedID: nil,
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.testName, func(t *testing.T) {
+			namespaceID, err := FindPublicNamespace(testCase.name, &config.CommandConfig{})
+			assert.Equal(t, testCase.expectedID, namespaceID, "Expected namespace ID to match")
+			assert.NoError(t, err, "Unexpected error from FindPrivateNamespace")
+		})
+	}
+}
+
+// Implements serviceDiscoveryClient interface
```

Comment:
I don't see how this is more difficult to maintain than our normal go mock code. Let me try to explain this by quickly writing in pseudocode how `TestFindPrivateNamespace` might look if it was written using the way that we normally write tests.

```
namespaceData := map[string]servicediscovery.Namespace{
// define a bunch of test data like I did in this PR
}

zoneData := map[string]route53.HostedZone{
// define a bunch of test data like I did in this PR
}

// set up go mocks
mockSD := mock_servicediscovery.NewClient()
mockR53 := mock_route53.NewClient()

gomock.InOrder(
    mockSD.EXPECT().ListNamespacePages(gomock.Any()).Return(namespaceData).Do(func (input) {
      // use .Do() to test the input
    }),
    // We need to list out every single API call that will be made to describe each namespace
    // This makes things tied to the implementation and harder to think about, because we have to figure out exactly what calls will be made
    mockR53.Expect().GetNamespace(gomock.Any()).Return(namespace1),
    mockR53.Expect().GetNamespace(gomock.Any()).Return(namespace2),
    // Note that we have to figure out which order the calls will be made in- this is because there is no way (that I know of) in gomock to use the argument to .GetNamespace() to figure out exactly which namespace to return- again making this way more complicated than the way I have done things.
   ....
   mockSD.Expect().GetHostedZone(gomock.Any()).Return(zone1),
   ....
)
```

As you can see, this certainly no less painful than what I have written. Personally, I find this more painful and more annoying- that's why I came up with the idea used in this PR.